rolling-shutter/p2pmsg/messages.go

```diff
@@ -0,0 +1,180 @@
+package p2pmsg
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+
+	shcrypto "github.com/shutter-network/shutter/shlib/shcrypto"
+
+	"github.com/shutter-network/rolling-shutter/rolling-shutter/keyper/kprtopics"
+	"github.com/shutter-network/rolling-shutter/rolling-shutter/medley/epochid"
+)
+
+// All messages to be used in the P2P Gossip have to be included in this slice,
+// otherwise they won't be known to the marshaling layer.
+var messageTypes = []P2PMessage{
+	// Keyper messages
+	new(DecryptionKey),
+	new(DecryptionTrigger),
+	new(DecryptionKeyShare),
+	new(EonPublicKey),
+}
+
+var topicToProtoName = make(map[string]protoreflect.FullName)
+
+func init() {
+	for _, mess := range messageTypes {
+		registerP2PMessage(mess)
+	}
+}
+
+// Instead of using an envelope for the unmarshalling,
+// we simply map one protobuf message type 1 to 1 to a Gossip topic.
+func registerP2PMessage(mess P2PMessage) {
+	messageTypeName := mess.ProtoReflect().Type().Descriptor().FullName()
+	topic := mess.Topic()
+
+	if val, exists := topicToProtoName[topic]; exists {
+		if val != messageTypeName {
+			err := errors.Errorf("Topic '%s' already has message type <%s> registered. Registering %s failed", topic, val, messageTypeName)
+			panic(err)
+		}
+	}
+	topicToProtoName[topic] = messageTypeName
+}
+
+// P2PMessage can be send via the p2p protocol.
+type P2PMessage interface {
```

Comment:
we could rename this now to `p2pmsg.Message`

Reply:
I'll probably do this in another commit